File: array/704.cpp
// divide and conquer (二分法)
// https://leetcode.cn/problems/binary-search/description/

#include <iostream>
#include <vector>  

using namespace std;

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int left = 0, right = nums.size()-1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] > target){
                right = mid - 1;
            }else if (nums[mid] < target){
                left = mid + 1;
            }else{
                return mid;
            }
        }
        return -1;
    }
};

int main() {
    vector<int> nums = {-1,0,3,5,9,12};
    int target = 9;
    Solution solution;
    int result = solution.search(nums, target);
    cout << result << endl;
    return 0;
}

File: array/704.py
from typing import List

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1
        while (left <= right):
            mid = left + (right - left) // 2
            if nums[mid] > target:
                right = mid - 1
            elif nums[mid] < target:
                left = mid + 1
            else:
                return mid
        return -1
        

if __name__ == '__main__':
    nums = [4,5,6,7,0,1,2]
    target = 0
    s = Solution()
    print(s.search(nums, target))